core/iwasm/aot/arch/aot_reloc_riscv.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2021 XiaoMi Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#include "aot_reloc.h"
+
+#define R_RISCV_CALL 18
+
+static SymbolMap target_sym_map[] = { REG_COMMON_SYMBOLS };
+
+void
+get_current_target(char *target_buf, uint32 target_buf_size)
+{
+    snprintf(target_buf, target_buf_size, "riscv");
+}
+
+uint32
+get_plt_item_size()
+{
+    return 16;
+}
+
+SymbolMap *
+get_target_symbol_map(uint32 *sym_num)
+{
+    *sym_num = sizeof(target_sym_map) / sizeof(SymbolMap);
+    return target_sym_map;
+}
+
+uint32
+get_plt_table_size()
+{
+    return get_plt_item_size() * (sizeof(target_sym_map) / sizeof(SymbolMap));
+}
+
+void
+init_plt_table(uint8 *plt)
+{}
+
+bool
+apply_relocation(AOTModule *module,
+                 uint8 *target_section_addr,
+                 uint32 target_section_size,
+                 uint64 reloc_offset,
+                 uint64 reloc_addend,
+                 uint32 reloc_type,
+                 void *symbol_addr,
+                 int32 symbol_index,
+                 char *error_buf,
+                 uint32 error_buf_size)
+{
+    uint8 *P = target_section_addr + reloc_offset;
+    uint32 hi20, lo12;
+
+    hi20 = (unsigned long)symbol_addr & 0xfffff000;
+    lo12 = ((unsigned long)symbol_addr & 0xfff) << 20;
+
+    switch (reloc_type) {
+        case R_RISCV_CALL:
+            /* lui  t0, hi20 */
+            *(uint32 *)(P + 0) = (uint32)hi20 | 0x2b7;
```

Comment:
Potential bugs here since we use absolute address to jump, it works fine on rv32, but on rv64 platform, target address may exceed the 32bit (4GB) address space.

Please take a look @wenyongh

Reply:
Yes, should add check, like R_X86_64_PC32/R_X86_64_32/R_X86_64_32S in aot_reloc_x86_64.c. Seems that here we should check whether (uint32)(uintptr_t)symbol_addr == (uintptr_t)symbol_addr?

If exceeds 4G, had better prompt using --size-level=1 if it fixes the issue.
And in aot_loader.c, try adding MMAP_MAP_32BIT for mmap:
```C
#if defined(BUILD_TARGET_X86_64) || defined(BUILD_TARGET_AMD_64) \
     || defined(BUILD_TARGET_RISCV64)
        /* aot code and data in x86_64 must be in range 0 to 2G due to
           relocation for R_X86_64_32/32S/PC32 */
        int map_flags = MMAP_MAP_32BIT;
#else
        int map_flags = MMAP_MAP_NONE;
#endif
```

Reply:
Done

Reply:
I prefer to implement all relocation types step by step since **R_RISCV_CALL** meet the requirements of our current workload.

Will this block the next release?

Reply:
It is OK, we had better test more cases, not sure whether your riscv hardware can run Linux system? If not, you can enable the QEMU to test them, ref to [Fedora QEMU](https://wiki.qemu.org/Documentation/Platforms/RISCV#Booting_64-bit_Fedora), and run it on Linux, the ABI is RISCV64_LP64D.